### PR TITLE
Move co-authored-by to end of messages

### PR DESCRIPTION
**Pull request by tomphp**
I find having them at the start gets in the way of the main title for the commit.

If there is any requirement to be able to have them at the start then I can look at adding a config option.

### CODE REVIEW

**Comment by PurpleBooth** [comment]
I think we can probably be more intelligent than that, we can detect if the git config has this key set, then if it does move it to the top, otherwise, leave it at the bottom

```toml
[commit]
	template = ~/.config/git/message
```

**Comment by PurpleBooth** [comment]
I suspect that the file with the commit message might already exist too... we might be able to do something to have different behavour based on that

**Comment by tomphp** [comment]
For some context, why should it be at the top if the template option is set?

**Comment by PurpleBooth** [comment]
If the template it set the commit file already has content in. This is what mine looks like

```
# Short (50 chars or less) summary of changes
#
# More detailed explanatory text, if necessary.  Wrap it to
# about 72 characters or so.  In some contexts, the first
# line is treated as the subject of an email and the rest of
# the text as the body.  The blank line separating the
# summary from the body is critical (unless you omit the body
# entirely); tools like rebase can get confused if you run
# the two together.
#
# Further paragraphs come after blank lines.
#
#   - Bullet points are okay, too
#
#   - Typically a hyphen or asterisk is used for the bullet,
#     preceded by a single space, with blank lines in
#     between, but conventions vary here

```

We read this file add our tag

```
# Short (50 chars or less) summary of changes
#
# More detailed explanatory text, if necessary.  Wrap it to
# about 72 characters or so.  In some contexts, the first
# line is treated as the subject of an email and the rest of
# the text as the body.  The blank line separating the
# summary from the body is critical (unless you omit the body
# entirely); tools like rebase can get confused if you run
# the two together.
#
# Further paragraphs come after blank lines.
#
#   - Bullet points are okay, too
#
#   - Typically a hyphen or asterisk is used for the bullet,
#     preceded by a single space, with blank lines in
#     between, but conventions vary here

Co-Authored-By: Annie Hexamle <annie@example.com>
```

**Comment by tomphp** [comment]
Maybe a cleverer approach would be to put it as the last line of uncommented message?

**Comment by PurpleBooth** [comment]
Yeah, I think that's probably the best approach

**Comment by PurpleBooth** [comment]
This will be fine for now though, like and #11 I think is the real "cause" of this, git is pretty complex and the tests don't really model it's behaviour accurately enough

**Comment by PurpleBooth** [comment]
Feel free just to merge this in whenever you're ready

**Comment by tomphp** [comment]
I'll have a crack at finding the right place to the message in the PR now anyway. Will take a look at #11 at another time.

**Comment by PurpleBooth** [comment]
Ah yeah, like don't worry about #11, I might try and do it if I find the motiviation

**Comment by PurpleBooth** [comment]
Oh, commented out text in commit messages isn't so straight forward btw

```
    Commands such as commit and tag that let you edit messages consider a line that begins with this character commented, and removes them after the editor returns (default #).

    If set to "auto", git-commit would select a character that is not the beginning character of any line in existing commit messages.

```

<https://git-scm.com/docs/git-config#Documentation/git-config.txt-corecommentChar>

**Comment by tomphp** [comment]
The comment char bit is not much of a problem but the "auto" bit is confusing, once it selects a character, how are you supposed to know what it is?

**Comment by PurpleBooth** [comment]
Yeah, I guess the idea is that nothing is commented, like a None type?

**Comment by tomphp** [comment]
I've had a bit of a go at making it a bit cleverer (sorry the PR got so long). A few notes about the PR:

- It puts the `Co-authored-by` trailer before the last comment in the message.
- It has `#` hardcoded as the commit message but it is parameterised in `CommitMessage` in preparation to read it from the config.
- `CommitMessage` has been moved to `pb-commit-message-lints/src/lib/commit_message.rs`. This is probably not where it should finally live but extracting it to it's own file felt like a good start.

**Comment by PurpleBooth** [comment]
Looks great as always @tomphp

File: pb-commit-message-lints/src/lints/lib/commit_message.rs [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+    #[must_use]
+    pub fn get_trailer(&self, trailer: &str) -> Vec<&str> {
+        self.contents
+            .lines()
+            .filter(|line: &&str| CommitMessage::line_has_trailer(trailer, line))
+            .collect::<Vec<_>>()
+    }
+
+    #[must_use]
+    pub fn add_trailer(&self, trailer: &str) -> Self {
```
Comment (PurpleBooth) [inline pb-commit-message-lints/src/lints/lib/commit_message.rs:35]: Can you talk me through the intention of this?

**Comment by PurpleBooth** [comment]
LGTM, mash that merge when you want

File: pb-commit-message-lints/src/lints/lib/commit_message.rs (hunk 1 above)
Comment (tomphp) [reply pb-commit-message-lints/src/lints/lib/commit_message.rs:35]: It takes the trailer string and it returns a new `CommitMessage` with the contents updated to include the new trailer in the right place.

It does this by splitting the message into 2 parts, the main message content and the trailing comment. It them puts them back together with the trailer line in-between.

Have I answered your question or have I missed what you were asking?

Comment (PurpleBooth) [reply pb-commit-message-lints/src/lints/lib/commit_message.rs:35]: Nah, I was just wondering about if it was used outside tests mostly

**Comment by PurpleBooth** [comment]
Merged! Thanks

File: pb-commit-message-lints/src/lints/lib/commit_message.rs (hunk 1 above)
Comment (tomphp) [reply pb-commit-message-lints/src/lints/lib/commit_message.rs:35]: It's used here https://github.com/PurpleBooth/pb-git-hooks/blob/master/pb-prepare-commit-msg/src/main.rs#L92

**Pull request merged**